### ISSUE TITLE
how to distinguish different client tcp traffic

### ISSUE THREAD

**Issue by joshua0x** [buger/goreplay#946]
> sudo ./goreplay --input-raw :8000 --output-http="http://example:8001"

according to codes , when multiple client send packets to 8000, every time pcap return tcp datas, below code is called , msg.Data maybe not commplete http packet ? and how to distinguish different client tcp traffic ?

<img width="709" alt="图片" src="https://user-images.githubusercontent.com/44366103/122914449-3cde8e80-d38d-11eb-9ac8-ded502797e3f.png">

**Comment by buger** [comment]
Hi! 

Can you explain your use-case? 
So basically you want to replay requests from the same user (same TCP session) using the same HTTP client?

**Comment by joshua0x** [comment]
>  So basically you want to replay requests from the same user (same TCP session) using the same HTTP client?

true, i want to use pcap capture tcp packets and get complete http request  to replay to endpoint ,  accroding to raw_input ，

>    sudo ./goreplay --input-raw :8000 --output-http="http://example:8001"  
in this case,  i guess  different client tcp sessions are interweaved ? ,

**Comment by buger** [comment]
I'm still quite confused
>  i want to use pcap capture tcp packets and get complete http request to replay to endpoint , accroding to raw_input ，

Essentially that how it works by design. It combines multiple packets together, reorder them if needed, and prepares HTTP message, which later can be either saved to file or directly replayed to another HTTP server. Multiple clients sending traffic to 8000 will work just fine, since it is a separate TCP streams and they will not interfere with each other. 

> in this case, i guess different client tcp sessions are interweaved ?

By default, GoRepay will not make difference between requests send in different TCP sessions. If I receive 10 messages from 10 different clients, while replaying requests via HTTP output, it will randomly distribute them to HTTP clients. However in [GoReplay PRO](https://goreplay.org/pro.html) you can have a mode where it will understand TCP sessions, and it will properly emulate them: if each client initiated 1 connection and sent 2 http messages during it, it will replay them exactly the same way in PRO version.

Hope it makes sense!

**Comment by joshua0x** [comment]
i want to know how goreplay define a tcp session  ? ,how  does a tcp session is assembly from original tcp messages ? 

> By default, GoRepay will not make difference between requests send in different TCP sessions. If I receive 10 messages from 10 different clients, while replaying requests via HTTP output, it will randomly distribute them to HTTP clients. However in [GoReplay PRO](https://goreplay.org/pro.html) you can have a mode where it will understand TCP sessions, and it will properly emulate them: if each client initiated 1 connection and sent 2 http messages during it, it will replay them exactly the same way in PRO version.

**Comment by joshua0x** [comment]
> i want to know how goreplay define a tcp session ? ,how does a tcp session is assembly from original tcp messages ?
> 
> > By default, GoRepay will not make difference between requests send in different TCP sessions. If I receive 10 messages from 10 different clients, while replaying requests via HTTP output, it will randomly distribute them to HTTP clients. However in [GoReplay PRO](https://goreplay.org/pro.html) you can have a mode where it will understand TCP sessions, and it will properly emulate them: if each client initiated 1 connection and sent 2 http messages during it, it will replay them exactly the same way in PRO version.

is it  possible one http req consist multiple tcp sessions ?

**Comment by buger** [comment]
In this case, i'm talking in standard Networking terms. 
TCP session, all the interactions between client and server from opening till closing connection.
In HTTP world it is common to have long opened sessions using keep-alive to keep it open  https://en.wikipedia.org/wiki/HTTP_persistent_connection
One TCP session can contain multiple HTTP requests, and one HTTP request can be sent only within the SAME TCP session.

**Comment by joshua0x** [comment]
> In this case, i'm talking in standard Networking terms.
> TCP session, all the interactions between client and server from opening till closing connection.
> In HTTP world it is common to have long opened sessions using keep-alive to keep it open https://en.wikipedia.org/wiki/HTTP_persistent_connection
> One TCP session can contain multiple HTTP requests, and one HTTP request can be sent only within the SAME TCP session.

thankyou ,so only when tcp closed ,pcap send data out once ? and the packet may contain multiple http req  ?

**Comment by buger** [comment]
No, it’s the stream, it receive data dynamically on the go. There’s is almost no delay between point in time when server received http request and GoReplay captured it.
